interpreter/types/types.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TypeProvider specifies functions for creating new object instances and for
+// resolving enum values by name.
+package types
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/ptypes/duration"
+	"github.com/golang/protobuf/ptypes/struct"
+	tspb "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/google/cel-go/interpreter/types/aspects"
+)
+
+// Declare the type instances and their human readable names.
+var (
+	TypeType      Type = &exprType{name: "type"}
+	NullType      Type = &exprType{name: "null"}
+	BoolType      Type = &exprType{name: "bool"}
+	IntType       Type = &exprType{name: "int"}
+	UintType      Type = &exprType{name: "uint"}
+	DoubleType    Type = &exprType{name: "double"}
+	StringType    Type = &exprType{name: "string"}
+	BytesType     Type = &exprType{name: "bytes"}
+	MapType       Type = &exprType{name: "map"}
+	ListType      Type = &exprType{name: "list"}
+	DurationType  Type = &exprType{name: "google.protobuf.Duration"}
+	TimestampType Type = &exprType{name: "google.protobuf.Timestamp"}
+	DynType       Type = &exprType{name: "dyn", isDyn: true}
+	// TODO: handle registration of abstract types, currently hard-coded.
+)
+
+// Types may be compared with each other, have a name, and indicate whether
+// the underlying type is dynamic.
+type Type interface {
+	aspects.Equaler
+	Name() string
+	IsDyn() bool
+}
+
+// MessageType produces a new Type instance for a qualified message name.
+func MessageType(name string) Type {
+	if name == TimestampType.Name() {
+		return TimestampType
+	} else if name == DurationType.Name() {
```

Comment:
It would be more idiomatic to drop the `else`.

Reply:
Done